`lv_textarea_delete_char`: GCC LTO warns `-Wstringop-overread` on `lv_strlen(ta->pwd_tmp)` (possible NULL dereference in pwd_mode)

### LVGL version

master (`d8ed8bc9e9f05e1182fe137cdb67d7f2de037bd9`)

### Platform

- LVGL: master (commit d8ed8bc9e9f05e1182fe137cdb67d7f2de037bd9)
- Compiler: GCC
- Build: Release with LTO enabled (-O2 -flto)
- OS: Linux
- Toolchain version: g++ (Ubuntu 15.2.0-16ubuntu1) 15.2.0


### What happened?

## Description

When building with LTO (e.g. `-O2 -flto`), GCC emits the following warning at link time:

```
In function 'lv_strlen',
    inlined from 'lv_textarea_delete_char' at lv_textarea.c:269:47,
    inlined from 'lv_textarea_delete_char' at lv_textarea.c:237:6:
lv_string_builtin.c:186:14: warning: '__builtin_strlen' reading 1 or more bytes
    from a region of size 0 [-Wstringop-overread]
  186 |     while(str[i]) i++;
lto1: note: source object is likely at address zero
```


### How to reproduce?

## Location

`src/widgets/textarea/lv_textarea.c`, inside the `pwd_mode` branch of `lv_textarea_delete_char`:

```c
if(ta->pwd_mode) {
    lv_text_cut(ta->pwd_tmp, ta->cursor.pos - 1, 1);

    ta->pwd_tmp = lv_realloc(ta->pwd_tmp, lv_strlen(ta->pwd_tmp) + 1);  // line 269
    LV_ASSERT_MALLOC(ta->pwd_tmp);
    if(ta->pwd_tmp == NULL) return;
}
```

## Analysis

GCC's LTO interprocedural analysis determines that `ta->pwd_tmp` can be at address zero (NULL) when `lv_strlen` is called on line 269. The NULL guard (`LV_ASSERT_MALLOC`) is placed *after* the `lv_strlen` call, which is too late.

A NULL check on `ta->pwd_tmp` before `lv_strlen` (or before `lv_text_cut`) would resolve this warning and guard against the potential dereference.

--- a/src/widgets/textarea/lv_textarea.c
+++ b/src/widgets/textarea/lv_textarea.c
@@ -219,9 +219,10 @@ void lv_textarea_add_text(lv_obj_t * obj, const char * txt)
 
     if(ta->pwd_mode) {
         size_t realloc_size = lv_strlen(ta->pwd_tmp) + lv_strlen(txt) + 1;
-        ta->pwd_tmp = lv_realloc(ta->pwd_tmp, realloc_size);
-        LV_ASSERT_MALLOC(ta->pwd_tmp);
-        if(ta->pwd_tmp == NULL) return;
+        char * new_buf = lv_realloc(ta->pwd_tmp, realloc_size);
+        LV_ASSERT_MALLOC(new_buf);
+        if(!new_buf) return;
+        ta->pwd_tmp = new_buf;
 
         lv_text_ins(ta->pwd_tmp, ta->cursor.pos, txt);
 
@@ -267,9 +268,10 @@ void lv_textarea_delete_char(lv_obj_t * obj)
     if(ta->pwd_mode) {
         lv_text_cut(ta->pwd_tmp, ta->cursor.pos - 1, 1);
 
-        ta->pwd_tmp = lv_realloc(ta->pwd_tmp, lv_strlen(ta->pwd_tmp) + 1);
-        LV_ASSERT_MALLOC(ta->pwd_tmp);
-        if(ta->pwd_tmp == NULL) return;
+        char * new_buf = lv_realloc(ta->pwd_tmp, lv_strlen(ta->pwd_tmp) + 1);
+        LV_ASSERT_MALLOC(new_buf);
+        if(!new_buf) return;
+        ta->pwd_tmp = new_buf;
     }
 
     /*Move the cursor to the place of the deleted character*/
@@ -389,7 +391,6 @@ void lv_textarea_set_password_mode(lv_obj_t * obj, bool en)
     lv_textarea_t * ta = (lv_textarea_t *)obj;
     if(ta->pwd_mode == en) return;
 
-    ta->pwd_mode = en ? 1U : 0U;
     /*Pwd mode is now enabled*/
     if(en) {
         char * txt = lv_label_get_text(ta->label);
@@ -398,12 +399,14 @@ void lv_textarea_set_password_mode(lv_obj_t * obj, bool en)
         LV_ASSERT_MALLOC(ta->pwd_tmp);
         if(ta->pwd_tmp == NULL) return;
 
+        ta->pwd_mode = 1U;
         pwd_char_hider(obj);
 
         lv_textarea_clear_selection(obj);
     }
     /*Pwd mode is now disabled*/
     else {
+        ta->pwd_mode = 0U;
         lv_textarea_clear_selection(obj);
         lv_label_set_text(ta->label, ta->pwd_tmp);
         lv_free(ta->pwd_tmp);

--- a/tests/src/test_cases/widgets/test_textarea.c
+++ b/tests/src/test_cases/widgets/test_textarea.c
@@ -428,6 +428,40 @@ void test_textarea_password_mode(void)
     TEST_ASSERT_EQUAL_SCREENSHOT("textarea_normal_mode.png");
 }
 
+/* In password mode, adding text must go through the realloc path of
+ * lv_textarea_add_text() so that pwd_tmp stays valid. */
+void test_textarea_add_text_in_password_mode(void)
+{
+    lv_textarea_set_password_mode(textarea, true);
+    lv_textarea_add_text(textarea, "Hello");
+    TEST_ASSERT_EQUAL_STRING("Hello", lv_textarea_get_text(textarea));
+
+    /*Adding more text should keep pwd_tmp in sync*/
+    lv_textarea_add_text(textarea, " World");
+    TEST_ASSERT_EQUAL_STRING("Hello World", lv_textarea_get_text(textarea));
+
+    lv_textarea_set_password_mode(textarea, false);
+    TEST_ASSERT_EQUAL_STRING("Hello World", lv_textarea_get_text(textarea));
+}
+
+/* In password mode, deleting a char must go through the realloc path of
+ * lv_textarea_delete_char() so that pwd_tmp stays valid. */
+void test_textarea_delete_char_in_password_mode(void)
+{
+    lv_textarea_set_text(textarea, "Hello World");
+    lv_textarea_set_password_mode(textarea, true);
+    lv_textarea_set_cursor_pos(textarea, 11);
+
+    lv_textarea_delete_char(textarea);
+    TEST_ASSERT_EQUAL_STRING("Hello Worl", lv_textarea_get_text(textarea));
+
+    lv_textarea_delete_char(textarea);
+    TEST_ASSERT_EQUAL_STRING("Hello Wor", lv_textarea_get_text(textarea));
+
+    lv_textarea_set_password_mode(textarea, false);
+    TEST_ASSERT_EQUAL_STRING("Hello Wor", lv_textarea_get_text(textarea));
+}
+
 void test_textarea_password_mode_hide_char(void)
 {
     lv_textarea_set_one_line(textarea, false);

